File: PAAKAT/LIBPLEPP/Tests/DUMMY/dummy.f90
program main
  implicit none
  include 'mpif.h'

  ! 0.0. 
  integer        :: error  = -1
  integer        :: gcomm, grank = -1, gsize = -1
  integer        :: lcomm, lrank = -1, lsize = -1

  integer        :: plepp_dim, dummy, timeSteps, itime, ii  
  character(64)  :: app_type, namei, namej

  real(8) :: send(2), recv(2) 

  logical :: test = .False. 

  ! 1.0. mpi init 
  lcomm   = mpi_comm_null
  gcomm   = MPI_COMM_WORLD
  call MPI_Init(error)
  call MPI_Comm_rank(gcomm, grank, error);
  call MPI_Comm_size(gcomm, gsize, error);

  ! 2.0. init_plepp
  plepp_dim  = 0   
  app_type   = "PLEPP"
  namei      = "ALYA"
  namej      = "HEMELB"
  lcomm      = gcomm  

  call plepp_create(trim(app_type), len_trim(app_type), &
                    trim(namei),    len_trim(namei),    &
                    trim(namej),    len_trim(namej),    &
                    plepp_dim, lcomm)
  call plepp_test() 

  ! 3.0. test_plepp     
  dummy     = 3
  timeSteps = -1  
  call plepp_sendrecv_int(dummy, 1, timeSteps, 1, 0) 

  recv = huge(1.0) 
  send = (/2.0,1.0/)
  do itime = 1,timeSteps 
    do ii = 1,size(send)  
      call plepp_exchange( send(ii), recv(ii), 0)
    enddo  
  enddo 

  test = all( recv==(/-1.0,-2.0/) )
  if(.not.test) then
    print *, "Test ERROR : recv==(/-1.0,-2.0/) !!" 
    call exit(0)
  endif 

  call MPI_Comm_rank(lcomm, lrank, error);
  if(lrank==0)  print *, "       ["//trim(namei)//"] recv:", recv

  call MPI_Finalize(error)

end program

File: PAAKAT/LIBPLEPP/Tests/DUMMY/dummy.cxx
/*
  2019NOV14. BSC, BARCELONA, SPAIN 
  Migue Zavala 
*/
#include <iostream>
#include <vector>
#include <map>
#include <mpi.h>
#include <cmath>
#include <assert.h>
#include <algorithm>
#include <sstream>
#include <iomanip>
#include <fstream>
#include <limits>
#include <cctype> // tolower 

#include "plepp.hpp"  
#include "insitu_py.hpp"

std::string GetLowers(std::string in)
{
  std::string out(in); 
  std::transform(out.begin(), out.end(), out.begin(), [](char c){ return std::tolower(c); });
  return out; 
}


int main(int argc, char** argv)
{
  //---------------------------------------------------------------------||---//
  assert(argc==2);
  std::string IAM = argv[1];


  //--------------------------------------------------------------| PLEPP |---//
  MPI_Init(NULL, NULL);

  HemelbCoupling *hemelbCoupling = NULL; 
  if(IAM.find("ALYA"  ) != std::string::npos) hemelbCoupling = new HemelbCoupling("PLEPP","ALYA",  "HEMELB", 0); 
  if(IAM.find("HEMELB") != std::string::npos) hemelbCoupling = new HemelbCoupling("PLEPP","HEMELB","ALYA"  , 0);
  assert(hemelbCoupling);

  MPI_Comm global_comm = MPI_COMM_WORLD; 
  MPI_Comm  local_comm = hemelbCoupling->Init( global_comm );  

  int far_root = hemelbCoupling->Test( hemelbCoupling->LRoot() );
  assert( (far_root==0)||(far_root==2) ); 

#ifdef WITHALYA  
#else     
  //---------------------------------------------------------------------||---//
  ofstream myfile;
  if( hemelbCoupling->LRoot() )  
  {
    myfile.open("iam_"+GetLowers(IAM) + ".dat");
    myfile<<"## itime recv";
    myfile<< scientific << showpoint;
    myfile<< setprecision( 10 );
    myfile<<" \n";
  }
  //hemelbCoupling->GBarrier();


  //----------------------------------| SimulationMaster::RunSimulation() |---//  
  int dummy     =  3;    
  int timeSteps = hemelbCoupling->SimulationSetTotalTimeSteps( dummy ); 


  //-----------------------------------------| LocalPropertyOutput::Write |---//
  for(int iSteps=0; iSteps<timeSteps; iSteps++)
  {
    std::vector<double> recv;  
    std::vector<double> send = {-1.0,-2.0}; 
    for(auto s: send) recv.push_back( hemelbCoupling->Exchange(s,0) ); 

    if( hemelbCoupling->LRoot() ) 
    { 
      myfile<<" "<< iSteps;
      for(auto r: recv) myfile<<" "<< r; 
      myfile<<" \n"; 
    }
  }

 
  //---------------------------------------------------------------------||---//
  if( hemelbCoupling->LRoot() )
  {
    std::cout<<"\t["<< IAM <<"] OK!! \n"; 
    myfile.close();
  }
#endif 

  MPI_Finalize();
  return 0;
}
//=======================================================================||===//
//=======================================================================||===//
